ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/RelOptHiveTable.java

```diff
@@ -275,103 +279,107 @@ public boolean isKey(ImmutableBitSet columns) {
+          parentTableQualifiedName.add(parentTableName);
+          qualifiedName = TableName.getDbTable(
+              parentDatabaseName, parentTableName);
+        } else {
+          parentTableQualifiedName.add(parentTableName);
+          qualifiedName = parentTableName;
+        }
+        Table parentTab = getTable(qualifiedName);
+        if (parentTab == null) {
+          LOG.error("Table for primary key not found: "
+              + "databaseName: " + parentDatabaseName + ", "
               + "tableName: " + parentTableName);
-        return ImmutableList.of();
-      }
-      ImmutableList.Builder<IntPair> keys = ImmutableList.builder();
-      for (ForeignKeyCol fkCol : fkCols) {
-        int fkPos;
-        for (fkPos = 0; fkPos < rowType.getFieldNames().size(); fkPos++) {
-          String fkColName = rowType.getFieldNames().get(fkPos);
-          if (fkColName.equals(fkCol.childColName)) {
-            break;
-          }
+          return ImmutableList.of();
         }
-        int pkPos;
-        for (pkPos = 0; pkPos < parentTab.getAllCols().size(); pkPos++) {
-          String pkColName = parentTab.getAllCols().get(pkPos).getName();
-          if (pkColName.equals(fkCol.parentColName)) {
-            break;
+        ImmutableList.Builder<IntPair> keys = ImmutableList.builder();
+        for (ForeignKeyCol fkCol : fkCols) {
+          int fkPos;
+          for (fkPos = 0; fkPos < rowType.getFieldNames().size(); fkPos++) {
+            String fkColName = rowType.getFieldNames().get(fkPos);
+            if (fkColName.equals(fkCol.childColName)) {
+              break;
+            }
           }
+          int pkPos;
+          for (pkPos = 0; pkPos < parentTab.getAllCols().size(); pkPos++) {
+            String pkColName = parentTab.getAllCols().get(pkPos).getName();
+            if (pkColName.equals(fkCol.parentColName)) {
+              break;
+            }
+          }
+          if (fkPos == rowType.getFieldNames().size()
+              || pkPos == parentTab.getAllCols().size()) {
+            LOG.error("Column for foreign key definition " + fkCol + " not found");
+            return ImmutableList.of();
+          }
+          keys.add(IntPair.of(fkPos, pkPos));
         }
-        if (fkPos == rowType.getFieldNames().size()
-            || pkPos == parentTab.getAllCols().size()) {
-          LOG.error("Column for foreign key definition " + fkCol + " not found");
-          return ImmutableList.of();
-        }
-        keys.add(IntPair.of(fkPos, pkPos));
+        builder.add(RelReferentialConstraintImpl.of(qualifiedTblName,
+            parentTableQualifiedName, keys.build()));
       }
-      builder.add(RelReferentialConstraintImpl.of(foreignKeyTableQualifiedName,
-              parentTableQualifiedName, keys.build()));
     }
     return builder.build();
   }
 
+  private Table getTable(String tableName) {
+    if (!tablesCache.containsKey(tableName)) {
+      try {
+        Table table = db.getTable(tableName);
```

Comment:
Null check for db? `HiveMaterializedViewsRegistry.java` is passing NULL for db.

Reply:
This is the qualified table name (see L1226 in Hive.java), not sure what you mean with NULL for db?

Reply:
This will never be called for materialized views, since currently you cannot define constraints on them. Not sure if you mean that?

Reply:
L:393 in `HiveMaterializedViewsRegistry` creates new `RelOptHiveTable` by passing NULL for `db`.  `getTable` here is using `db` to call `getTable`.  I meant it should be checked if `db` is null or not before calling `getTable` on it

Reply:
I prefer that it fails instead of ignoring the retrieval silently, as it should never be null unless you know what you are doing, i.e. materialized views. I can add a comment in the table creation in ```HiveMaterializedViewsRegistry```.